### PR TITLE
Throw ArgumentException for invalid uses of global members in Expressions.

### LINKED ISSUE #12943: System.Linq.Expressions factories throw NullReferenceException with global fields and global methods
Consider the following test class I put together of all instances. I will submit a PR fixing these bugs if it you guys approve, so please let me know. Here I've used RefEmit to produce the global fields/methods, as C# doesn't support them

```
public class GlobalMethodGlobalField
{
    [Fact]
    public void AndAlso_GlobalMethod_ThrowsNullReferenceException()
    {
        MethodInfo method = GlobalMethod(typeof(int), new Type[] { typeof(int), typeof(int) });
        Assert.Throws<NullReferenceException>(() => Expression.AndAlso(Expression.Constant(1), Expression.Constant(2), method));
    }

    [Fact]
    public void Bind_GlobalMethod_ThrowsNullReferenceException()
    {
        Assert.Throws<NullReferenceException>(() => Expression.Bind(GlobalMethod(typeof(int), new Type[0]), Expression.Constant("abc")));
    }

    [Fact]
    public void Default_TypeBuilder_CreateLambda_ThrowsNotSupportedException()
    {
        AssemblyBuilder assemblyBuilder = AssemblyBuilder.DefineDynamicAssembly(new AssemblyName("Assembly"), AssemblyBuilderAccess.Run);
        ModuleBuilder moduleBuilder = assemblyBuilder.DefineDynamicModule("Module");
        TypeBuilder typeBuilder = moduleBuilder.DefineType("Type");

        DefaultExpression defaultExpression = Expression.Default(typeBuilder.AsType());
        Assert.Throws<NotSupportedException>(() => Expression.Lambda(defaultExpression));

        LambdaExpression genericLambda = Expression.Lambda<Func<object>>(defaultExpression);
        Assert.Null(genericLambda.Compile());
    }

    [Fact]
    public void ListBind_GlobalMethod_ThrowsNullReferenceException()
    {
        Assert.Throws<NullReferenceException>(() => Expression.ListBind(GlobalMethod(typeof(int), new Type[0]), new ElementInit[0]));
        Assert.Throws<NullReferenceException>(() => Expression.ListBind(GlobalMethod(typeof(int), new Type[0]), (IEnumerable<ElementInit>)new ElementInit[0]));
    }

    [Fact]
    public void MemberBind_GlobalMethod_ThrowsNullReferenceException()
    {
        Assert.Throws<NullReferenceException>(() => Expression.MemberBind(GlobalMethod(typeof(int), new Type[0]), new MemberBinding[0]));
        Assert.Throws<NullReferenceException>(() => Expression.MemberBind(GlobalMethod(typeof(int), new Type[0]), (IEnumerable<MemberBinding>)new MemberBinding[0]));
    }

    [Fact]
    public void MemberBind_GlobalFieldInBindings_ThrowsNullReferenceException()
    {
        MemberBinding[] bindings = new MemberBinding[] { Expression.MemberBind(GlobalField()) };
        Assert.Throws<NullReferenceException>(() => Expression.MemberBind(GlobalField(), bindings));
        Assert.Throws<NullReferenceException>(() => Expression.MemberBind(GlobalField(), (IEnumerable<MemberBinding>)bindings));

    }

    [Fact]
    public void New_GlobalMethodInMembers_ThrowsNullReferenceException()
    {
        ConstructorInfo constructor = typeof(ClassWithConstructor).GetConstructor(new Type[] { typeof(int) });
        Expression[] arguments = new Expression[] { Expression.Constant(5) };
        MemberInfo[] members = new MemberInfo[] { GlobalMethod(typeof(int), new Type[0]) };
        Assert.Throws<NullReferenceException>(() => Expression.New(constructor, arguments, members));
    }

    [Fact]
    public void New_GlobalFieldInMembers_ThrowsNullReferenceException()
    {
        ConstructorInfo constructor = typeof(ClassWithConstructor).GetConstructor(new Type[] { typeof(int) });
        Expression[] arguments = new Expression[] { Expression.Constant(5) };
        MemberInfo[] members = new MemberInfo[] { GlobalField() };
        Assert.Throws<NullReferenceException>(() => Expression.New(constructor, arguments, members));
    }

    [Fact]
    public void OrElse_GlobalMethod_ThrowsNullReferenceException()
    {
        MethodInfo method = GlobalMethod(typeof(int), new Type[] { typeof(int), typeof(int) });
        Assert.Throws<NullReferenceException>(() => Expression.OrElse(Expression.Constant(1), Expression.Constant(2), method));
    }

    private class ClassWithConstructor
    {
        public ClassWithConstructor(int i) { }
    }

    private static MethodInfo GlobalMethod(Type returnType, Type[] parameterTypes)
    {
        AssemblyBuilder assembly = AssemblyBuilder.DefineDynamicAssembly(new AssemblyName("ABC"), AssemblyBuilderAccess.Run);
        ModuleBuilder module = assembly.DefineDynamicModule("Module");

        MethodBuilder globalMethod = module.DefineGlobalMethod("GlobalMethod", MethodAttributes.Public | MethodAttributes.Static, returnType, parameterTypes);
        globalMethod.GetILGenerator().Emit(OpCodes.Ret);

        module.CreateGlobalFunctions();
        return module.GetMethod(globalMethod.Name);
    }

    private static FieldInfo GlobalField()
    {
        AssemblyBuilder assembly = AssemblyBuilder.DefineDynamicAssembly(new AssemblyName("ABC"), AssemblyBuilderAccess.Run);
        ModuleBuilder module = assembly.DefineDynamicModule("Module");

        FieldBuilder fieldBuilder = module.DefineInitializedData("Name", new byte[1], FieldAttributes.Public);

        module.CreateGlobalFunctions();
        return module.GetField(fieldBuilder.Name);
    }
}
```

@bartdesmet @VSadov @JonHanna

### PR DESCRIPTION
**Pull request by JonHanna**
Several methods of `Expression` will throw `NullReferenceException` if called with a `MethodInfo` or `FieldInfo` for a global method or field. These exceptions are all thrown from validation of the (null) `DeclaringType`, so fix these validation methods to throw the same exception they would throw for a non-null but inappropriate `DeclaringType`.

Using a global field with a `MemberBinding` does not throw, but any attempt to use that binding will similarly throw `NullReferenceException`. Change to throw an `ArgumentException` immediately.

Fixes #12943

### CODE REVIEW

File: src/System.Linq.Expressions/src/Resources/Strings.resx [hunk 1]
```diff
@@ -366,6 +366,9 @@
   <data name="NotAMemberOfType" xml:space="preserve">
     <value>'{0}' is not a member of type '{1}'</value>
   </data>
+  <data name="NotAMemberOfAnyType" xml:space="preserve">
+    <value>'{0}' is not a member of any </value>
```
Comment (nil4) [inline src/System.Linq.Expressions/src/Resources/Strings.resx]: last word missing (*type*)?

Comment (JonHanna) [inline src/System.Linq.Expressions/src/Resources/Strings.resx]: Oops. Thanks.

File: src/Common/src/System/Dynamic/Utils/TypeUtils.cs [hunk 2]
```diff
@@ -10,7 +10,7 @@ internal static partial class TypeUtils
     {
         public static bool AreEquivalent(Type t1, Type t2)
         {
-            return t1.IsEquivalentTo(t2);
+            return t1 != null && t1.IsEquivalentTo(t2);
```
Comment (VSadov) [inline src/Common/src/System/Dynamic/Utils/TypeUtils.cs]: Type implements "==" operator. We just want a null check.
Would it be better to do something like:   `return t1?.IsEquivalentTo(t2) ?? false`

File: src/System.Linq.Expressions/src/System/Dynamic/Utils/TypeUtils.cs [hunk 3]
```diff
@@ -180,6 +180,10 @@ public static bool IsValidInstanceType(MemberInfo member, Type instanceType)
             {
                 return true;
             }
+            if (targetType == null)
```
Comment (VSadov) [inline src/System.Linq.Expressions/src/System/Dynamic/Utils/TypeUtils.cs]: `it ((object)targetType == null)`   -   just to make it cheaper by not using the "==" operator on Type

File: src/System.Linq.Expressions/src/System/Linq/Expressions/BinaryExpression.cs [hunk 4]
```diff
@@ -824,8 +824,13 @@ private static void ValidateUserDefinedConditionalLogicOperator(ExpressionType n
             {
                 left = left.GetNonNullableType();
             }
-            MethodInfo opTrue = TypeUtils.GetBooleanOperator(method.DeclaringType, "op_True");
-            MethodInfo opFalse = TypeUtils.GetBooleanOperator(method.DeclaringType, "op_False");
+            Type declaringType = method.DeclaringType;
+            if (declaringType == null)
```
Comment (VSadov) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/BinaryExpression.cs]: same thing about  Type and "=="

Actually, It seems MethodInfo also has "==" implemented. 
Perhaps it is worth going through and turn trivial null checks with Type/MemberInfo and the like into object equality or ?. operators. As a separate change of course.

**Comment by VSadov** [comment]
LGTM

**Pull request merged**